src/magicbox/QueriesProcessor.ts

```diff
@@ -0,0 +1,91 @@
+import { defaults } from 'underscore';
+
+export interface IQueriesProcessorOptions {
+  timeout: number;
+}
+
+export enum QueryProcessResultStatus {
+  Finished = 'finished',
+  TimedOut = 'timedout',
+  Overriden = 'overriden'
+}
+
+export type QueryProcessResult<ItemType> =
+  | {
+      status: QueryProcessResultStatus.Finished | QueryProcessResultStatus.TimedOut;
+      items: ItemType[];
+    }
+  | {
+      status: QueryProcessResultStatus.Overriden;
+    };
+
+export class QueriesProcessor<ItemType> {
+  private static async streamQueries<ItemType>(queries: Promise<ItemType[]>[], output: ItemType[]) {
+    await Promise.all(queries.map(query => query.then(items => output.push(...items))));
```

Comment:
Why does it return stuff in "output" ? 
Why not return the value like a normal function ?

Reply:
This is so that the results of `streamQueries` can be obtained before it finishes processing every query. This is used in `processQueries` so that it can be timed out or overriden before this `Promise.all` resolves and still return the results accumulated so far.

Would a private property be better? (e.g., `this.processedItems.push(...item)`)
Or do you have any better solution in mind? This is a function I wasn't too sure how to clean up.